OpenSim/Simulation/Test/testRealizeDynamics.cpp

```diff
@@ -0,0 +1,101 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+#include <stdint.h>
+#include <OpenSim/Simulation/Manager/Manager.h>
+#include <OpenSim/Simulation/Control/ControlSetController.h>
+#include <OpenSim/Simulation/Model/Model.h>
+#include <OpenSim/Common/LoadOpenSimLibrary.h>
+#include <OpenSim/Auxiliary/auxiliaryTestFunctions.h>
+
+using namespace OpenSim;
+using namespace std;
+
+//==============================================================================
+// testRealizeDynamics tests that the results produced in every stage are 
+// repeatable - this test is designed primarily to make sure that there are no 
+// memory races during realizeDynamics that can cause time-dependent results 
+// with regards to thread timing.
+//==============================================================================
+void testRealizeDynamics(const string& modelFile);
+
+static const int MAX_N_TRIES = 100;
+
+int main()
+{
+    try {
+        LoadOpenSimLibrary("osimActuators");
+        testRealizeDynamics("arm26.osim");
+    }
+    catch (const Exception& e) {
+        cout << "testRealizeDynamics failed: ";
+        e.print(cout);
+        return 1;
+    }
+    catch (const std::exception& e) {
+        cout << "testRealizeDynamics failed: " << e.what() << endl;
+        return 1;
+    }
+    cout << "Done" << endl;
+    return 0;
+}
+
+//==============================================================================
+// Test Cases
+//==============================================================================
+void testRealizeDynamics(const string& modelFile)
+{
+    using namespace SimTK;
+    using namespace std;
+
+    //==========================================================================
+    // Setup OpenSim model
+    Model model(modelFile);
+
+    ControlSetController* controller = new ControlSetController();
+    controller->setControlSetFileName("arm26_StaticOptimization_controls.xml");
+    model.addController(controller);
+
+    //==========================================================================
+    // Test Accuracy of Dynamics
+    
+    //Realize to Stage::Dynamics once to obtain a Dyanmics "answer"
+    State& state = model.initSystem();
+    model.getSystem().realize(state, Stage::Dynamics);
+    const MultibodySystem& mbs = model.getMultibodySystem();
+    Vector_<SpatialVec>&   rigidBodyForcesAnswer = mbs.updRigidBodyForces(state,
+                                                               Stage::Dynamics);
+    
+    string answerString = rigidBodyForcesAnswer.toString(); 
+    //Constantly realize to Stage::Dynamics to make sure that our answers are
+    //consistent
+    for(int x = 0; x < 100; x++)
```

Comment:
We have a convention of preincrementing: https://github.com/opensim-org/opensim-core/blob/master/CONTRIBUTING.md#always-use-pre-increment-and-pre-decrement-operators-when-you-have-a-choice